// babel-plugin-dva-hmr/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (_ref) {
  var t = _ref.types;

  var cache = {};

  function getImportRequirePath(identifierName, scope) {
    if (scope.hasBinding(identifierName)) {
      var binding = scope.bindings[identifierName];
      if (binding) {
        var parent = binding.path.parent;

        if (t.isImportDeclaration(parent)) {
          return parent.source.value;
        } else if (t.isVariableDeclaration(parent)) {
          var declarator = findDeclarator(parent.declarations, identifierName);
          if (declarator && isRequire(declarator.init)) {
            return declarator.init.arguments[0].value;
          }
        }
      }
    }
    return null;
  }

  function isDvaCallExpression(node, scope) {
    return t.isCallExpression(node) && t.isIdentifier(node.callee) && getImportRequirePath(node.callee.name, scope) === 'dva';
  }

  function isDvaInstance(identifierName, scope) {
    if (scope.hasBinding(identifierName)) {
      var binding = scope.bindings[identifierName];
      if (binding) {
        var parent = binding.path.parent;
        if (t.isVariableDeclaration(parent)) {
          var declarator = findDeclarator(parent.declarations, identifierName);
          if (declarator && isDvaCallExpression(declarator.init, scope)) {
            return true;
          }
        }
      }
    }
    return false;
  }

  function isRouterCall(node, scope) {
    if (!t.isMemberExpression(node)) return false;
    var object = node.object;
    var property = node.property;

    return t.isIdentifier(property) && property.name === 'router' && t.isIdentifier(object) && isDvaInstance(object.name, scope);
  }

  function isRequire(node) {
    return t.isCallExpression(node) && t.isIdentifier(node.callee) && node.callee.name === 'require';
  }

  function findDeclarator(declarations, identifier) {
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = declarations[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var d = _step.value;

        if (t.isIdentifier(d.id) && d.id.name === identifier) {
          return d;
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  }

  function getRouterPath(node, scope, filename, opts) {
    switch (node.type) {
      case 'CallExpression':
        if (t.isLiteral(node.arguments[0])) {
          return node.arguments[0].value;
        }
        break;
      case 'Identifier':
        var path = getImportRequirePath(node.name, scope);
        if (path) {
          return path;
        }
        break;
      default:
        break;
    }
    !opts.quiet && console.warn('[babel-plugin-dva-hmr][WARN] can\'t get router path in ' + filename);
  }

  return {
    visitor: {
      CallExpression: function CallExpression(path, _ref2) {
        var opts = _ref2.opts;
        var filename = path.hub.file.opts.filename;

        if (cache[filename]) return;
        var _path$node = path.node;
        var callee = _path$node.callee;
        var args = _path$node.arguments;

        if (isRouterCall(callee, path.scope)) {
          var routerPath = getRouterPath(args[0], path.scope, filename, opts);
          if (routerPath) {
            cache[filename] = true;
            !opts.quiet && console.info('[babel-plugin-dva-hmr][INFO] got routerPath ' + routerPath);
            path.parentPath.replaceWithSourceString(getHmrString(callee.object.name, routerPath, opts.container));
          }
        }
      }
    }
  };
};

var _path = require('path');

function getHmrString(appName, routerPath) {
  var container = arguments.length <= 2 || arguments[2] === undefined ? '#root' : arguments[2];

  return '\n(function() {\n  // Generated by babel-plugin-dva-hmr\n  console.log(\'[HMR] inited with babel-plugin-dva-hmr\');\n  ' + appName + '.router(require(\'' + routerPath + '\'));\n  ' + appName + '.use({\n    onHmr(render) {\n      if (module.hot) {\n        const renderNormally = render;\n        const renderException = (error) => {\n          const RedBox = require(\'redbox-react\');\n          ReactDOM.render(React.createElement(RedBox, { error: error }), document.querySelector(\'' + container + '\'));\n        };\n        const newRender = (router) => {\n          try {\n            renderNormally(router);\n          } catch (error) {\n            console.error(\'error\', error);\n            renderException(error);\n          }\n        };\n        module.hot.accept(\'' + routerPath + '\', () => {\n          const router = require(\'' + routerPath + '\');\n          newRender(router);\n        });\n      }\n    },  \n  });\n})()\n    ';
}

module.exports = exports['default'];
