// react-stack-grid/lib/animations/easings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

// http://easings.net/
var linear = exports.linear = "linear";
var easeIn = exports.easeIn = "ease-in";
var easeOut = exports.easeOut = "ease-out";
var easeInOut = exports.easeInOut = "ease-in-out";
var sineIn = exports.sineIn = "cubic-bezier(0.47, 0, 0.745, 0.715)";
var sineOut = exports.sineOut = "cubic-bezier(0.39, 0.575, 0.565, 1)";
var sineInOut = exports.sineInOut = "cubic-bezier(0.445, 0.05, 0.55, 0.95)";
var quadIn = exports.quadIn = "cubic-bezier(0.55, 0.085, 0.68, 0.53)";
var quadOut = exports.quadOut = "cubic-bezier(0.25, 0.46, 0.45, 0.94)";
var quadInOut = exports.quadInOut = "cubic-bezier(0.455, 0.03, 0.515, 0.955)";
var cubicIn = exports.cubicIn = "cubic-bezier(0.55, 0.055, 0.675, 0.19)";
var cubicOut = exports.cubicOut = "cubic-bezier(0.215, 0.61, 0.355, 1)";
var cubicInOut = exports.cubicInOut = "cubic-bezier(0.645, 0.045, 0.355, 1)";
var quartIn = exports.quartIn = "cubic-bezier(0.895, 0.03, 0.685, 0.22)";
var quartOut = exports.quartOut = "cubic-bezier(0.165, 0.84, 0.44, 1)";
var quartInOut = exports.quartInOut = "cubic-bezier(0.77, 0, 0.175, 1)";
var quintIn = exports.quintIn = "cubic-bezier(0.755, 0.05, 0.855, 0.06)";
var quintOut = exports.quintOut = "cubic-bezier(0.23, 1, 0.32, 1)";
var quintInOut = exports.quintInOut = "cubic-bezier(0.86, 0, 0.07, 1)";
var expoIn = exports.expoIn = "cubic-bezier(0.95, 0.05, 0.795, 0.035)";
var expoOut = exports.expoOut = "cubic-bezier(0.19, 1, 0.22, 1)";
var expoInOut = exports.expoInOut = "cubic-bezier(1, 0, 0, 1)";
var circIn = exports.circIn = "cubic-bezier(0.6, 0.04, 0.98, 0.335)";
var circOut = exports.circOut = "cubic-bezier(0.075, 0.82, 0.165, 1)";
var circInOut = exports.circInOut = "cubic-bezier(0.785, 0.135, 0.15, 0.86)";
var backIn = exports.backIn = "cubic-bezier(0.6, -0.28, 0.735, 0.045)";
var backOut = exports.backOut = "cubic-bezier(0.175, 0.885, 0.32, 1.275)";
var backInOut = exports.backInOut = "cubic-bezier(0.68, -0.55, 0.265, 1.55)";
