// react-stack-grid/lib/components/StackGrid.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridInline = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _reactDom = require("react-dom");

var _reactDom2 = _interopRequireDefault(_reactDom);

var _reactAddonsTransitionGroup = require("react-addons-transition-group");

var _reactAddonsTransitionGroup2 = _interopRequireDefault(_reactAddonsTransitionGroup);

var _reactSizeme = require("react-sizeme");

var _reactSizeme2 = _interopRequireDefault(_reactSizeme);

var _shallowequal = require("shallowequal");

var _shallowequal2 = _interopRequireDefault(_shallowequal);

var _throttleDebounce = require("throttle-debounce");

var _exenv = require("exenv");

var _exenv2 = _interopRequireDefault(_exenv);

var _invariant = require("invariant");

var _invariant2 = _interopRequireDefault(_invariant);

var _GridItem = require("./GridItem");

var _GridItem2 = _interopRequireDefault(_GridItem);

var _styleHelper = require("../utils/style-helper");

var _easings = require("../animations/easings");

var easings = _interopRequireWildcard(_easings);

var _transitions = require("../animations/transitions/");

var transitions = _interopRequireWildcard(_transitions);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var imagesLoaded = _exenv2.default.canUseDOM ? require("imagesloaded") : null;

var isNumber = function isNumber(v) {
  return typeof v === "number" && isFinite(v);
};
var isPercentageNumber = function isPercentageNumber(v) {
  return typeof v === "string" && /^\d+(\.\d+)?%$/.test(v);
};

var createArray = function createArray(v, l) {
  var array = [];
  for (var i = 0; i < l; i++) {
    array.push(v);
  }return array;
};

var propTypes = {
  children: _react.PropTypes.node,
  className: _react.PropTypes.string,
  style: _react.PropTypes.object,
  component: _react.PropTypes.string,
  columnWidth: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]).isRequired,
  gutterWidth: _react.PropTypes.number,
  gutterHeight: _react.PropTypes.number,
  duration: _react.PropTypes.number,
  easing: _react.PropTypes.string,
  appearDelay: _react.PropTypes.number,
  appear: _react.PropTypes.func,
  appeared: _react.PropTypes.func,
  enter: _react.PropTypes.func,
  entered: _react.PropTypes.func,
  leaved: _react.PropTypes.func,
  units: _react.PropTypes.shape({
    length: _react.PropTypes.string,
    angle: _react.PropTypes.string
  }),
  monitorImagesLoaded: _react.PropTypes.bool,
  vendorPrefix: _react.PropTypes.bool,
  userAgent: _react.PropTypes.string
};

var GridInline = exports.GridInline = function (_Component) {
  _inherits(GridInline, _Component);

  function GridInline(props) {
    _classCallCheck(this, GridInline);

    var _this = _possibleConstructorReturn(this, (GridInline.__proto__ || Object.getPrototypeOf(GridInline)).call(this, props));

    _this.handleItemMounted = function (item) {
      var key = item.props.itemKey;

      _this.items[key] = item;
      _this.updateLayout(_this.props);

      if (_this.props.monitorImagesLoaded && typeof imagesLoaded === "function") {
        var node = _reactDom2.default.findDOMNode(item);
        var imgLoad = imagesLoaded(node);

        imgLoad.on("always", function () {
          _this.debouncedUpdateLayout(_this.props);
        });

        _this.imgLoad[key] = imgLoad;
      }
    };

    _this.handleItemUnmount = function (item) {
      var key = item.props.itemKey;


      if (_this.items.hasOwnProperty(key)) {
        delete _this.items[key];
      }

      if (_this.imgLoad.hasOwnProperty(key)) {
        _this.imgLoad[key].off("always");
        delete _this.imgLoad[key];
      }
    };

    _this.items = {};
    _this.imgLoad = {};
    _this.debouncedUpdateLayout = (0, _throttleDebounce.debounce)(20, _this.updateLayout);
    _this.state = _this.doLayout(props);
    return _this;
  }

  _createClass(GridInline, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.updateLayout(this.props);
    }
  }, {
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      if (!(0, _shallowequal2.default)(nextProps, this.props)) {
        this.updateLayout(nextProps);
      }
    }
  }, {
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState) {
      return !(0, _shallowequal2.default)(nextProps, this.props) || !(0, _shallowequal2.default)(nextState, this.state);
    }
  }, {
    key: "getItemHeight",
    value: function getItemHeight(item) {
      if (item.key && this.items.hasOwnProperty(item.key)) {
        var _component = this.items[item.key];
        var el = _reactDom2.default.findDOMNode(_component);
        var candidate = [el.scrollHeight, el.clientHeight, el.offsetHeight, 0].filter(isNumber);

        return Math.max.apply(Math, _toConsumableArray(candidate));
      }

      return 0;
    }
  }, {
    key: "getColumnLengthAndWidth",
    value: function getColumnLengthAndWidth(width, value, gutter) {
      if (isNumber(value)) {
        var _columnWidth = parseFloat(value, 10);

        return [Math.floor((width - (width / _columnWidth - 1) * gutter) / _columnWidth), _columnWidth];
      } else if (isPercentageNumber(value)) {
        var _columnWidth2 = width * (parseFloat(value, 10) / 100);
        var maxColumn = Math.floor(width / _columnWidth2);

        return [maxColumn, _columnWidth2 - gutter * (maxColumn - 1)];
      }

      (0, _invariant2.default)(false, "Should be columnWidth is a number or percentage string.");
    }
  }, {
    key: "doLayout",
    value: function doLayout(props) {
      return _exenv2.default.canUseDOM && props.size.width !== 0 ? this.doLayoutForClient(props) : this.doLayoutForSSR(props);
    }
  }, {
    key: "doLayoutForClient",
    value: function doLayoutForClient(props) {
      var _this2 = this;

      var containerWidth = props.size.width,
          rawColumnWidth = props.columnWidth,
          gutterWidth = props.gutterWidth,
          gutterHeight = props.gutterHeight;


      var childArray = _react2.default.Children.toArray(props.children);

      var _getColumnLengthAndWi = this.getColumnLengthAndWidth(containerWidth, rawColumnWidth, gutterWidth),
          _getColumnLengthAndWi2 = _slicedToArray(_getColumnLengthAndWi, 2),
          maxColumn = _getColumnLengthAndWi2[0],
          columnWidth = _getColumnLengthAndWi2[1];

      var columnHeights = createArray(0, maxColumn);

      var rects = childArray.map(function (child) {
        var column = columnHeights.indexOf(Math.min.apply(Math, _toConsumableArray(columnHeights)));
        var height = _this2.getItemHeight(child);
        var left = column * columnWidth + column * gutterWidth;
        var top = columnHeights[column];

        columnHeights[column] += Math.round(height) + gutterHeight;

        return { top: top, left: left, width: columnWidth, height: height };
      });

      var width = maxColumn * columnWidth + (maxColumn - 1) * gutterWidth;
      var height = Math.max.apply(Math, _toConsumableArray(columnHeights)) - gutterHeight;
      var finalRects = rects.map(function (o) {
        return _extends({}, o, {
          left: o.left + (containerWidth - width) / 2
        });
      });

      return { rects: finalRects, actualWidth: width, height: height, columnWidth: columnWidth };
    }
  }, {
    key: "doLayoutForSSR",
    value: function doLayoutForSSR(props) {
      return {
        rects: _react2.default.Children.toArray(props.children).map(function () {
          return {
            top: 0, left: 0, width: 0, height: 0
          };
        }),
        actualWidth: 0,
        height: 0,
        columnWidth: 0
      };
    }
  }, {
    key: "updateLayout",
    value: function updateLayout(props) {
      this.setState(this.doLayout(props));
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _props = this.props,
          gutterWidth = _props.gutterWidth,
          gutterHeight = _props.gutterHeight,
          rawColumnWidth = _props.columnWidth,
          monitorImagesLoaded = _props.monitorImagesLoaded,
          className = _props.className,
          style = _props.style,
          size = _props.size,
          component = _props.component,
          children = _props.children,
          rest = _objectWithoutProperties(_props, ["gutterWidth", "gutterHeight", "columnWidth", "monitorImagesLoaded", "className", "style", "size", "component", "children"]);

      var _state = this.state,
          rects = _state.rects,
          actualWidth = _state.actualWidth,
          height = _state.height;

      var containerSize = {
        actualWidth: actualWidth,
        width: size.width == null ? 0 : size.width,
        height: height
      };
      var validChildren = _react2.default.Children.toArray(children).filter(function (child) {
        return (0, _react.isValidElement)(child);
      });

      /* eslint-disable no-return-assign */
      return _react2.default.createElement(
        _reactAddonsTransitionGroup2.default,
        {
          component: component,
          className: className,
          style: _extends({}, style, {
            position: "relative",
            transition: (0, _styleHelper.transition)(["height"], rest.duration, easings.easeOut),
            height: height
          })
        },
        validChildren.map(function (child, i) {
          return _react2.default.createElement(
            _GridItem2.default,
            _extends({}, rest, {
              index: i,
              key: child.key,
              itemKey: child.key,
              rect: rects[i],
              containerSize: containerSize,
              onMounted: _this3.handleItemMounted,
              onUnmount: _this3.handleItemUnmount
            }),
            child
          );
        })
      );
      /* eslint-enable no-return-assign */
    }
  }]);

  return GridInline;
}(_react.Component);

GridInline.propTypes = _extends({}, propTypes, {
  size: _react.PropTypes.shape({
    width: _react.PropTypes.number,
    height: _react.PropTypes.number
  })
});


var SizeAwareGridInline = (0, _reactSizeme2.default)({
  monitorWidth: true,
  monitorHeight: false
})(GridInline);

var StackGrid = function (_Component2) {
  _inherits(StackGrid, _Component2);

  function StackGrid() {
    _classCallCheck(this, StackGrid);

    return _possibleConstructorReturn(this, (StackGrid.__proto__ || Object.getPrototypeOf(StackGrid)).apply(this, arguments));
  }

  _createClass(StackGrid, [{
    key: "render",
    value: function render() {
      return _react2.default.createElement(SizeAwareGridInline, this.props);
    }
  }]);

  return StackGrid;
}(_react.Component);

StackGrid.propTypes = propTypes;
StackGrid.defaultProps = {
  style: {},
  component: "div",
  columnWidth: 150,
  gutterWidth: 5,
  gutterHeight: 5,
  duration: 480,
  easing: easings.quartOut,
  appearDelay: 30,
  appear: transitions.fadeUp.appear,
  appeared: transitions.fadeUp.appeared,
  enter: transitions.fadeUp.enter,
  entered: transitions.fadeUp.entered,
  leaved: transitions.fadeUp.leaved,
  units: { length: "px", angle: "deg" },
  monitorImagesLoaded: false,
  vendorPrefix: true,
  userAgent: null
};
exports.default = StackGrid;
